radix-engine/src/kernel/call_frame.rs

```diff
@@ -117,55 +120,67 @@ impl CallFrame {
         offset: SubstateOffset,
         flags: LockFlags,
     ) -> Result<LockHandle, RuntimeError> {
-        let location = self.get_node_location(node_id)?;
+        self.check_node_visibility(&node_id)?;
         if !(matches!(offset, SubstateOffset::KeyValueStore(..))
             || matches!(
                 offset,
                 SubstateOffset::NonFungibleStore(NonFungibleStoreOffset::Entry(..))
             ))
         {
             let substate_id = SubstateId(node_id, module_id, offset.clone());
-            match location {
-                RENodeLocation::Store => track
-                    .acquire_lock(substate_id, flags)
-                    .map_err(KernelError::TrackError),
-                RENodeLocation::Heap => {
-                    if flags.contains(LockFlags::UNMODIFIED_BASE) {
-                        Err(KernelError::TrackError(
-                            TrackError::LockUnmodifiedBaseOnNewSubstate(substate_id),
-                        ))
-                    } else {
-                        Ok(())
-                    }
+            if heap.contains_node(&node_id) {
```

Comment:
Not sure about this interface where we check heap first since location of node should be well known from the time of it's creation.